### PR TITLE
new unit tests in dark-gravity-wave-test added

### PR DESCRIPTION
**Pull request by abvgedeika**

### CODE REVIEW

File: test/dark-gravity-wave-test.js [hunk 1]
```diff
@@ -1,126 +1,286 @@
...
+            expect(result).to.equal(true);
+        });
+
+        it('should be invalid maxTarget = 0x207fffff', () => {
+            const highTargetInt = 0x207fffff;
+            const result = dgw.isValidTarget(highTargetInt, blocks);
+            expect(result).to.equal(false);
+        });
+
+        it('should be valid when num blocks < maxBlocks and maxTarget = 0x207fffff', () => {//TODO is it True?
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: yes as I understand if less than 25 blocks has been mined the maxtarget is valid.

File: test/dark-gravity-wave-test.js [hunk 2]
```diff
@@ -1,126 +1,286 @@
...
+            const highTargetInt = 0x207fffff;
+            const result = dgw.isValidTarget(highTargetInt, blocks.slice(0, 10));
+            expect(result).to.equal(true);// TODO change on false?
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: not sure what you mean by _change on false_?

File: test/dark-gravity-wave-test.js [hunk 3]
```diff
@@ -1,126 +1,286 @@
...
+
+        it('should throw error for allHeaders is {}', () => {
+            const highTargetBits = 0x208fffff;
+            expect(() => {
+                dgw.isValidTarget(highTargetBits, {});
+            }).to.throw('allHeaders.slice is not a function');
+        });
+
+        it('should throw error for allHeaders is string', () => {
+            // TODO not very good that we allow string although it's length < 25 and ignored in getTarget
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: yes agree.  maybe we should add some validation on the `allHeaders` param provided?

File: test/dark-gravity-wave-test.js [hunk 4]
```diff
@@ -1,126 +1,286 @@
...
+
+        it('should throw error when compact is boolean', () => {
+            expect(() => {
+                dgw.isValidTarget(true, blocks);
+            }).to.throw('Compact of type boolean not supported');
+        });
+
+        it('height values ignored in calculation in getTarget function', () => {
+            //TODO is it true?
+            //and we can pass any blocks height: invalid, duplicates, not ordered...?
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: yes I think that is fine.  Consensus rules do not consider height and its up to the dev to provide correct `allHeaders`.  Since DGW is stateless it cannot determine if the previous 25 headers provided is conforming to consensus based on earlier blocks.

File: test/dark-gravity-wave-test.js [hunk 5]
```diff
@@ -1,126 +1,286 @@
...
+            expect(result).to.equal(false);
+       });
+
+        it('should be invalid for blockTime=0', () => {
+            const highTargetBits = 0x1b0777d4; // 453474260
+            const result = dgw.isValidTarget(highTargetBits, blocks,  blockTime = 0);
+            expect(result).to.equal(false);
+        });
+
+        it('should be invalid for blockTime negative', () => {//TODO We should not allow negative blockTime
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: agree

File: test/dark-gravity-wave-test.js [hunk 6]
```diff
@@ -1,126 +1,286 @@
...
+            const highTargetBits = 0x1b0777d4; // 453474260
+            const result = dgw.isValidTarget(highTargetBits, blocks,  blockTime = -1);
+            expect(result).to.equal(true);//TODO change on true!
+        });
+
+        it('should be invalid when timestamp for some blocks grater then next one', () => {//TODO We should check if timestamp is grater then previous
+            const highTargetBits = 0x1b0777d4;
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: more or less similar reasoning as in line 209.  however we might consider notifying the dev with an exception if things  like this is picked up.  Then again for `timestamp` in specific I am guessing that there is edge cases where timestamps can be out of order and still valid - Im not really sure.  Leaning towards not checking these fields.

File: test/dark-gravity-wave-test.js [hunk 7]
```diff
@@ -1,126 +1,286 @@
...
+            blocks2.forEach(function(value){
+                if (Math.random() >= 0.3){
+                    value["timestamp"]+=2000
+                }
+            });
+            const result = dgw.isValidTarget(highTargetBits, blocks2);
+            expect(result).to.equal(true);//TODO change on false!
+        });
+
+        it('should be invalid when timestamp for some blocks grater then next one', () => {//TODO We should check if timestamp is grater then previous?
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: similar as per prev comment ^

File: test/dark-gravity-wave-test.js [hunk 8]
```diff
@@ -1,126 +1,286 @@
...
+                if (Math.random() >= 0.3){
+                    value["timestamp"]+=2000
+                }
+            });
+            const result = dgw.isValidTarget(highTargetBits, blocks2);
+            expect(result).to.equal(true);//TODO change on false!
+        });
+
+
+        it('should be invalid when reverse block array', () => {//TODO Should we check order of blocks?
```
Comment (pvrooyen) [inline test/dark-gravity-wave-test.js]: similar as per line 209

**Review by pvrooyen — APPROVED** [review]
utAck with comments

**Review by pvrooyen — DISMISSED** [review]
utAck with comments

**Review by pvrooyen — APPROVED** [review]
utAck
